src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/ActivityService.kt

```diff
@@ -228,17 +230,14 @@ class ActivityService(
     val courseActivity = activityRepository.findByIdOrNull(courseActivityId)
       ?: throw NotFoundException("Course Activity $courseActivityId not found")
 
-    if (courseActivity.scheduleEndDate != null && courseActivity.scheduleEndDate!! < date) {
-      throw BadDataException("Course Activity id ${courseActivity.courseActivityId} ended on ${courseActivity.scheduleEndDate}")
-    }
-
     with(courseActivity) {
-      scheduleEndDate = date
-      endComment?.run { commentText = endComment }
+      if (courseActivity.scheduleEndDate == null || courseActivity.scheduleEndDate!! > date) {
+        scheduleEndDate = date
+        endComment?.run { commentText = endComment }
+      }
     }
-    courseActivity.offenderProgramProfiles
+    offenderProgramProfileRepository.findByCourseActivityCourseActivityIdAndProgramStatusCode(courseActivityId, "ALLOC")
```

Comment:
Some activities have many thousands of ended allocations that we're not interested in and cause performance problems - so just get the active allocations.